File: 2516-take-k-of-each-character-from-left-and-right/2516-take-k-of-each-character-from-left-and-right.java
class Solution {
    public int takeCharacters(String s, int k) {
        int n = s.length();
        int a = 0, b = 0, c = 0;

        for (char ch : s.toCharArray()) {
            if (ch == 'a') a++;
            if (ch == 'b') b++;
            if (ch == 'c') c++;
        }

        if (a < k || b < k || c < k) return -1;

        int i = 0, j = 0;
        int ans = 0;

        while (j < n) {
            if (s.charAt(j) == 'a') a--;
            if (s.charAt(j) == 'b') b--;
            if (s.charAt(j) == 'c') c--;

            while (i <= j && (a < k || b < k || c < k)) {
                if (s.charAt(i) == 'a') a++;
                if (s.charAt(i) == 'b') b++;
                if (s.charAt(i) == 'c') c++;
                i++;
            }

            ans = Math.max(ans, j - i + 1);
            j++;
        }

        return n - ans;
    }
}

File: LeetCode/2516-take-k-of-each-character-from-left-and-right/2516-take-k-of-each-character-from-left-and-right.py
class Solution:
    def takeCharacters(self, s: str, k: int) -> int:
        n = len(s)
        a = b = c = 0

        for ch in s:
            if ch == 'a': a += 1
            if ch == 'b': b += 1
            if ch == 'c': c += 1

        if a < k or b < k or c < k:
            return -1

        i, j = 0, 0
        ans = 0

        while j < n:
            if s[j] == 'a': a -= 1
            if s[j] == 'b': b -= 1
            if s[j] == 'c': c -= 1

            while i <= j and (a < k or b < k or c < k):
                if s[i] == 'a': a += 1
                if s[i] == 'b': b += 1
                if s[i] == 'c': c += 1
                i += 1

            ans = max(ans, j - i + 1)
            j += 1

        return n - ans